README.md
<img src="./docs/imgs/pygimbal.png" width="100%">

# pygimbal

`Pygimbal` is an open source software developed in `Python`, designed for controlling gimbals using the `Mavlink protocol`. It allows users to control the gimbal's tilt, pan, and roll angles, as well as retrieve real-time angle values.

## Install

```
pip install pygimbal
```

## Getting Started

How to control gimbal via Mavlink with Python:

> Note: you need to give permission to gimbal uart port first.

```shell
sudo chmod a+wrx /dev/ttyUSB0
```

Sample code:


```python
import time
from pygimbal import control
from threading import Thread

master_gimbal = control.init_mav_connector('/dev/ttyUSB0', 1, 154, 115200)
system = master_gimbal.source_system
component = master_gimbal.source_component

maintain_connection_threading = Thread(target=control.maintain_connection, args=[master_gimbal], daemon=True)
maintain_connection_threading.start()

orientation_theading = Thread(target=control.get_orientation, args=[master_gimbal], daemon=True)
orientation_theading.start()

while True:
    try:
        control.control_gimbal(master_gimbal, system, component, tilt=45, roll=0, pan=45)
        time.sleep(2)
        control.control_gimbal(master_gimbal, system, component, tilt=0, roll=0, pan=0)
        time.sleep(2)
        control.control_gimbal(master_gimbal, system, component, tilt=45, roll=0, pan=-45)
        time.sleep(2)
        control.control_gimbal(master_gimbal, system, component, tilt=0, roll=0, pan=0)
        time.sleep(2)
    except KeyboardInterrupt:
        break
```

Output:

```
[ACTION]: Mavlink is connecting to /dev/ttyUSB0
[MESSAGE]: Successful Connection to /dev/ttyUSB0 1 154
[Control] Gimbal Pitch - Yaw 45 45
[ORIENTATION] Til, Pan, Roll:  0.1949998289346695 0.02197265625 0.023102451115846634
[ORIENTATION] Til, Pan, Roll:  0.19493775069713593 0.02197265625 0.02414652332663536
[ORIENTATION] Til, Pan, Roll:  0.19543787837028503 0.02197265625 0.02471373975276947
[ORIENTATION] Til, Pan, Roll:  0.19867371022701263 0.02197265625 0.0238236952573061
```

Result:

<div align='center'>
  <img src="./docs/imgs/result.gif" width="70%">
</div>

## Author

```
github: winter2897
mail: haiquantran2897@gmail.com
```

<!-- ## Licence

Pygimbal is made available under the permissive open source Apache 2.0 License.

Copyright 2023 Kirins Robotics, Inc. -->

CHANGELOG.md
# Changelog

## Version 1.0.0

### New Features:

* Add connect to gimbal via pymavlink `init_mav_connector`.
* Add control gimbal `control_gimbal`.
* Add maintain connection to gimbal `maintain_connection`.
* Add get orientation from gimbal `get_orientation`.


sample/sample_control.py
import time
from pygimbal import control
from threading import Thread

master_gimbal = control.init_mav_connector('/dev/ttyUSB0', 1, 154, 115200)
system = master_gimbal.source_system
component = master_gimbal.source_component

maintain_connection_threading = Thread(target=control.maintain_connection, args=[master_gimbal], daemon=True)
maintain_connection_threading.start()

orientation_theading = Thread(target=control.get_orientation, args=[master_gimbal], daemon=True)
orientation_theading.start()

while True:
    try:
        control.control_gimbal(master_gimbal, system, component, tilt=45, roll=0, pan=45)
        time.sleep(2)
        control.control_gimbal(master_gimbal, system, component, tilt=0, roll=0, pan=0)
        time.sleep(2)
        control.control_gimbal(master_gimbal, system, component, tilt=45, roll=0, pan=-45)
        time.sleep(2)
        control.control_gimbal(master_gimbal, system, component, tilt=0, roll=0, pan=0)
        time.sleep(2)
    except KeyboardInterrupt:
        break

pygimbal/control.py
#!/bin/bash
# author='winter2897'

import os
os.environ['MAVLINK20'] = '1'
os.environ['MAVLINK_DIALECT'] = 'ardupilotmega'

import time
from threading import Thread
from pymavlink import mavutil

def init_mav_connector(device, source_system, source_component, baud=115200):

    print('[ACTION]: Mavlink is connecting to', device)
    master = mavutil.mavlink_connection(device, baud, source_system, source_component, autoreconnect=True)
    master.mav.heartbeat_send(2,0,0,0,0,3)
    master.wait_heartbeat()
    print('[MESSAGE]: Successful Connection to', device, master.source_system, master.source_component)

    return master

def control_gimbal(master,target, source, tilt, roll, pan):

    master.mav.command_long_send(
            target,
            source,
            mavutil.mavlink.MAV_CMD_DO_MOUNT_CONTROL, 
            1,
            tilt,
            roll,
            pan,
            0, 0, 0,
            mavutil.mavlink.MAV_MOUNT_MODE_MAVLINK_TARGETING)
    
    print('[Control] Gimbal Pitch - Yaw', tilt, pan)
            
def maintain_connection(master):
    while True:
        master.mav.heartbeat_send(2,0,0,0,0,3)
        time.sleep(0.5)

def get_orientation(master):

    while True:
        msg = master.recv_match()
        if msg:
            if msg.get_type() == 'MOUNT_ORIENTATION':
                tilt = msg.pitch
                pan = msg.yaw
                roll = msg.roll
                print('[ORIENTATION] Til, Pan, Roll: ', tilt, pan, roll)
        time.sleep(0.01)

if __name__=="__main__":

    master_gimbal = init_mav_connector('/dev/ttyUSB0', 1, 154, 115200)
    system = master_gimbal.source_system
    component = master_gimbal.source_component

    maintain_connection_threading = Thread(target=maintain_connection, args=[master_gimbal], daemon=True)
    maintain_connection_threading.start()

    orientation_theading = Thread(target=get_orientation, args=[master_gimbal], daemon=True)
    orientation_theading.start()

    while True:
        control_gimbal(master_gimbal, system, component, tilt=45, roll=0, pan=45)
        time.sleep(2)
        control_gimbal(master_gimbal, system, component, tilt=0, roll=0, pan=0)
        time.sleep(2)
        control_gimbal(master_gimbal, system, component, tilt=45, roll=0, pan=-45)
        time.sleep(2)
        control_gimbal(master_gimbal, system, component, tilt=0, roll=0, pan=0)
        time.sleep(2)

